File: src/services/championChallengeService.ts
import { championChallengeApiService, ExecutionResponse } from './championChallengeApiService';
import type { ChampionChallengeExecution, NodeMetric } from '../types/championChallenge';

// Helper to safely parse JSON (handles both string and already-parsed objects)
function safeParse(value: any): any {
  if (!value) return {};
  if (typeof value === 'string') {
    try {
      return JSON.parse(value);
    } catch {
      return {};
    }
  }
  return value;
}

// Helper function to convert API response to frontend format
function mapApiToFrontend(apiResponse: ExecutionResponse): ChampionChallengeExecution {
  const championMetrics: NodeMetric[] = apiResponse.championMetrics.map(metric => ({
    id: metric.id,
    executionId: apiResponse.id,
    variant: 'champion' as const,
    nodeId: metric.nodeId,
    nodeName: metric.nodeName,
    nodeType: metric.nodeType,
    requestData: safeParse(metric.requestData),
    responseData: safeParse(metric.responseData),
    executionTimeMs: metric.executionTimeMs,
    status: metric.status.toLowerCase() as 'success' | 'error' | 'skipped',
    errorMessage: metric.errorMessage || undefined,
    startedAt: new Date(metric.startedAt),
    completedAt: metric.completedAt ? new Date(metric.completedAt) : undefined,
    metadata: safeParse(metric.metadata),
  }));

  const challengeMetrics: NodeMetric[] = apiResponse.challengeMetrics.map(metric => ({
    id: metric.id,
    executionId: apiResponse.id,
    variant: 'challenge' as const,
    nodeId: metric.nodeId,
    nodeName: metric.nodeName,
    nodeType: metric.nodeType,
    requestData: safeParse(metric.requestData),
    responseData: safeParse(metric.responseData),
    executionTimeMs: metric.executionTimeMs,
    status: metric.status.toLowerCase() as 'success' | 'error' | 'skipped',
    errorMessage: metric.errorMessage || undefined,
    startedAt: new Date(metric.startedAt),
    completedAt: metric.completedAt ? new Date(metric.completedAt) : undefined,
    metadata: safeParse(metric.metadata),
  }));

  return {
    id: apiResponse.id,
    name: apiResponse.name,
    description: apiResponse.description,
    championWorkflowId: apiResponse.championWorkflowId,
    challengeWorkflowId: apiResponse.challengeWorkflowId,
    requestPayload: safeParse(apiResponse.requestPayload),
    status: apiResponse.status.toLowerCase() as 'running' | 'completed' | 'failed',
    startedAt: new Date(apiResponse.startedAt),
    completedAt: apiResponse.completedAt ? new Date(apiResponse.completedAt) : undefined,
    createdAt: new Date(apiResponse.createdAt),
    metrics: {
      champion: championMetrics,
      challenge: challengeMetrics,
    },
  };
}

// Mock data generator matching API format
function generateMockExecution(
  name: string,
  description: string,
  championWorkflowId: string,
  challengeWorkflowId: string
): ChampionChallengeExecution {
  const id = `mock-${Date.now()}`;
  const nodes = [
    { id: 'start', name: 'StartEvent', type: 'Event' },
    { id: 'validate', name: 'ServiceTask_ValidateInput', type: 'ServiceTask' },
    { id: 'process', name: 'ServiceTask_ProcessData', type: 'ServiceTask' },
    { id: 'call', name: 'ServiceTask_CallAPI', type: 'ServiceTask' },
    { id: 'gateway', name: 'Gateway_CheckCondition', type: 'Gateway' },
    { id: 'handle', name: 'ServiceTask_HandleResult', type: 'ServiceTask' },
    { id: 'end', name: 'EndEvent', type: 'Event' },
  ];

  const generateMetrics = (variant: 'champion' | 'challenge'): NodeMetric[] => {
    const baseMultiplier = variant === 'champion' ? 1 : 0.8;
    return nodes.map((node, index) => {
      const execTime = Math.floor((Math.random() * 200 + 100) * baseMultiplier);
      const status = Math.random() > 0.95 ? 'error' : 'success';

      return {
        id: `${id}-${variant}-${index}`,
        executionId: id,
        variant,
        nodeId: node.id,
        nodeName: node.name,
        nodeType: node.type,
        requestData: { payload: 'test' },
        responseData: { result: status === 'success' ? 'success' : 'error' },
        executionTimeMs: execTime,
        status: status as 'success' | 'error',
        errorMessage: status === 'error' ? `Error in ${node.name}` : undefined,
        startedAt: new Date(Date.now() - execTime),
        completedAt: new Date(),
        metadata: {
          memoryUsed: 20 + Math.random() * 80,
          cpuUsage: 10 + Math.random() * 80,
        },
      };
    });
  };

  return {
    id,
    name,
    description,
    championWorkflowId,
    challengeWorkflowId,
    requestPayload: {},
    status: 'completed',
    startedAt: new Date(Date.now() - 5000),
    completedAt: new Date(),
    createdAt: new Date(Date.now() - 10000),
    metrics: {
      champion: generateMetrics('champion'),
      challenge: generateMetrics('challenge'),
    },
  };
}

function generateMockExecutions(): ChampionChallengeExecution[] {
  return [
    generateMockExecution(
      'Payment Flow Performance Test',
      'Testing payment processing speed',
      'payment-v1',
      'payment-v2'
    ),
    generateMockExecution(
      'Order Fulfillment Comparison',
      'Comparing order workflows',
      'order-v1',
      'order-v2'
    ),
  ];
}

export class ChampionChallengeService {
  private useBackend: boolean = true;
  private mockExecutions: ChampionChallengeExecution[] = generateMockExecutions();

  async executeComparison(
    championWorkflowId: string,
    challengeWorkflowId: string,
    requestPayload: any,
    name: string,
    description?: string
  ): Promise<ChampionChallengeExecution> {
    if (this.useBackend) {
      try {
        const apiResponse = await championChallengeApiService.createExecution(
          name,
          description || '',
          championWorkflowId,
          challengeWorkflowId,
          requestPayload
        );
        return mapApiToFrontend(apiResponse);
      } catch (error) {
        console.warn('Backend API failed, using mock data:', error);
        this.useBackend = false;
      }
    }

    const mockExecution = generateMockExecution(
      name,
      description || '',
      championWorkflowId,
      challengeWorkflowId
    );
    this.mockExecutions.unshift(mockExecution);
    await new Promise(resolve => setTimeout(resolve, 1000));
    return mockExecution;
  }

  async loadExecution(executionId: string): Promise<ChampionChallengeExecution | null> {
    if (this.useBackend) {
      try {
        const apiResponse = await championChallengeApiService.getExecution(executionId);
        return mapApiToFrontend(apiResponse);
      } catch (error) {
        console.warn('Backend API failed, using mock data:', error);
        this.useBackend = false;
      }
    }

    const mockExecution = this.mockExecutions.find(e => e.id === executionId);
    return mockExecution || null;
  }

  async listExecutions(): Promise<ChampionChallengeExecution[]> {
    if (this.useBackend) {
      try {
        const apiResponses = await championChallengeApiService.listExecutions();
        return apiResponses.map(mapApiToFrontend);
      } catch (error) {
        console.warn('Backend API failed, using mock data:', error);
        this.useBackend = false;
      }
    }

    return this.mockExecutions;
  }

  resetToBackend() {
    this.useBackend = true;
  }
}

export const championChallengeService = new ChampionChallengeService();

File: backend-complete/src/main/java/com/wellsfargo/workflow/championchallenge/service/ChampionChallengeService.java
package com.wellsfargo.workflow.championchallenge.service;

import com.wellsfargo.workflow.championchallenge.dto.*;
import com.wellsfargo.workflow.championchallenge.entity.*;
import com.wellsfargo.workflow.championchallenge.repository.*;
import com.wellsfargo.workflow.common.service.WorkflowExecutionService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class ChampionChallengeService {

    private final ChampionChallengeExecutionRepository executionRepository;
    private final ExecutionNodeMetricRepository metricRepository;
    private final WorkflowExecutionService workflowExecutionService;

    private static final List<String> SAMPLE_NODES = Arrays.asList(
            "StartEvent", "ServiceTask_ValidateInput", "ServiceTask_ProcessData",
            "ServiceTask_CallAPI", "Gateway_CheckCondition", "ServiceTask_HandleResult",
            "EndEvent"
    );

    @Transactional
    public ExecutionResponse createExecution(ExecutionRequest request, String userId) {
        log.info("Creating execution: {}", request.getName());

        ChampionChallengeExecutionEntity execution = new ChampionChallengeExecutionEntity();
        execution.setName(request.getName());
        execution.setDescription(request.getDescription());
        execution.setChampionWorkflowId(request.getChampionWorkflowId());
        execution.setChallengeWorkflowId(request.getChallengeWorkflowId());
        execution.setRequestPayload(request.getRequestPayload());
        execution.setCreatedBy(userId);
        execution.setStatus(ChampionChallengeExecutionEntity.ExecutionStatus.RUNNING);
        execution.setStartedAt(LocalDateTime.now());

        execution = executionRepository.save(execution);

        long championTotal = executeWorkflow(execution, "CHAMPION", request.getChampionWorkflowId());
        long challengeTotal = executeWorkflow(execution, "CHALLENGE", request.getChallengeWorkflowId());

        execution.setTotalChampionTimeMs(championTotal);
        execution.setTotalChallengeTimeMs(challengeTotal);
        execution.setWinner(championTotal < challengeTotal ? "CHAMPION" : "CHALLENGE");
        execution.setStatus(ChampionChallengeExecutionEntity.ExecutionStatus.COMPLETED);
        execution.setCompletedAt(LocalDateTime.now());

        executionRepository.save(execution);

        return mapToResponse(execution);
    }

    private long executeWorkflow(ChampionChallengeExecutionEntity execution, String variant, String workflowId) {
        long totalTime = 0L;
        boolean isChallenge = "CHALLENGE".equals(variant);

        for (String nodeName : SAMPLE_NODES) {
            LocalDateTime start = LocalDateTime.now();
            long execTime;
            String status = "SUCCESS";
            String errorMessage = null;

            try {
                execTime = workflowExecutionService.simulateNodeExecution(nodeName, isChallenge);
            } catch (Exception e) {
                execTime = 0L;
                status = "ERROR";
                errorMessage = e.getMessage();
            }

            ExecutionNodeMetricEntity metric = new ExecutionNodeMetricEntity();
            metric.setExecution(execution);
            metric.setVariant(variant);
            metric.setNodeId(nodeName.toLowerCase());
            metric.setNodeName(nodeName);
            metric.setNodeType(nodeName.contains("Task") ? "ServiceTask" :
                    nodeName.contains("Gateway") ? "Gateway" : "Event");
            metric.setExecutionTimeMs(execTime);
            metric.setStatus(status);
            metric.setErrorMessage(errorMessage);
            metric.setStartedAt(start);
            metric.setCompletedAt(LocalDateTime.now());

            execution.addNodeMetric(metric);
            metricRepository.save(metric);

            totalTime += execTime;
        }

        return totalTime;
    }

    @Transactional(readOnly = true)
    public ExecutionResponse getExecution(String executionId) {
        ChampionChallengeExecutionEntity execution = executionRepository.findByIdWithMetrics(executionId)
                .orElseThrow(() -> new RuntimeException("Execution not found"));
        return mapToResponse(execution);
    }

    @Transactional(readOnly = true)
    public List<ExecutionResponse> listExecutions() {
        return executionRepository.findAllByOrderByCreatedAtDesc().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public AnalyticsResponse getAnalytics(String executionId) {
        ChampionChallengeExecutionEntity execution = executionRepository.findByIdWithMetrics(executionId)
                .orElseThrow(() -> new RuntimeException("Execution not found"));

        List<ExecutionNodeMetricEntity> championMetrics = execution.getNodeMetrics().stream()
                .filter(m -> "CHAMPION".equals(m.getVariant()))
                .collect(Collectors.toList());

        List<ExecutionNodeMetricEntity> challengeMetrics = execution.getNodeMetrics().stream()
                .filter(m -> "CHALLENGE".equals(m.getVariant()))
                .collect(Collectors.toList());

        double improvement = execution.getTotalChampionTimeMs() > 0
                ? ((execution.getTotalChampionTimeMs() - execution.getTotalChallengeTimeMs()) /
                (double) execution.getTotalChampionTimeMs()) * 100
                : 0.0;

        AnalyticsResponse.SummaryCards summaryCards = AnalyticsResponse.SummaryCards.builder()
                .championTime(execution.getTotalChampionTimeMs())
                .challengeTime(execution.getTotalChallengeTimeMs())
                .winner(execution.getWinner())
                .improvement(improvement)
                .build();

        List<AnalyticsResponse.ExecutionTimeData> executionTimeData = new ArrayList<>();
        for (int i = 0; i < championMetrics.size() && i < challengeMetrics.size(); i++) {
            executionTimeData.add(AnalyticsResponse.ExecutionTimeData.builder()
                    .node(championMetrics.get(i).getNodeName())
                    .champion(championMetrics.get(i).getExecutionTimeMs())
                    .challenge(challengeMetrics.get(i).getExecutionTimeMs())
                    .build());
        }

        List<AnalyticsResponse.PieData> pieData = Arrays.asList(
                AnalyticsResponse.PieData.builder().name("Champion").value(execution.getTotalChampionTimeMs()).build(),
                AnalyticsResponse.PieData.builder().name("Challenge").value(execution.getTotalChallengeTimeMs()).build()
        );

        long champSuccess = championMetrics.stream().filter(m -> "SUCCESS".equals(m.getStatus())).count();
        long challSuccess = challengeMetrics.stream().filter(m -> "SUCCESS".equals(m.getStatus())).count();

        List<AnalyticsResponse.SuccessRateData> successRateData = Arrays.asList(
                AnalyticsResponse.SuccessRateData.builder()
                        .variant("Champion")
                        .success(champSuccess)
                        .error(championMetrics.size() - champSuccess)
                        .build(),
                AnalyticsResponse.SuccessRateData.builder()
                        .variant("Challenge")
                        .success(challSuccess)
                        .error(challengeMetrics.size() - challSuccess)
                        .build()
        );

        List<AnalyticsResponse.CumulativeData> cumulativeData = new ArrayList<>();
        long champCumulative = 0L;
        long challCumulative = 0L;
        for (int i = 0; i < championMetrics.size() && i < challengeMetrics.size(); i++) {
            champCumulative += championMetrics.get(i).getExecutionTimeMs();
            challCumulative += challengeMetrics.get(i).getExecutionTimeMs();
            cumulativeData.add(AnalyticsResponse.CumulativeData.builder()
                    .node(championMetrics.get(i).getNodeName())
                    .championCumulative(champCumulative)
                    .challengeCumulative(challCumulative)
                    .build());
        }

        AnalyticsResponse.RadarData radarData = AnalyticsResponse.RadarData.builder()
                .metrics(Arrays.asList("Speed", "Efficiency", "Reliability", "Performance", "Quality"))
                .champion(Arrays.asList(85.0, 90.0, 95.0, 88.0, 92.0))
                .challenge(Arrays.asList(90.0, 85.0, 90.0, 92.0, 88.0))
                .build();

        List<AnalyticsResponse.PerformanceComparison> performanceComparison = Arrays.asList(
                AnalyticsResponse.PerformanceComparison.builder()
                        .metric("Avg Time").championValue(championMetrics.stream().mapToLong(ExecutionNodeMetricEntity::getExecutionTimeMs).average().orElse(0.0))
                        .challengeValue(challengeMetrics.stream().mapToLong(ExecutionNodeMetricEntity::getExecutionTimeMs).average().orElse(0.0)).build(),
                AnalyticsResponse.PerformanceComparison.builder()
                        .metric("Total Time").championValue(execution.getTotalChampionTimeMs().doubleValue())
                        .challengeValue(execution.getTotalChallengeTimeMs().doubleValue()).build()
        );

        AnalyticsResponse.DetailedStatistics detailedStatistics = AnalyticsResponse.DetailedStatistics.builder()
                .totalChampionNodes((long) championMetrics.size())
                .totalChallengeNodes((long) challengeMetrics.size())
                .championSuccess(champSuccess)
                .challengeSuccess(challSuccess)
                .championSuccessRate(championMetrics.size() > 0 ? (champSuccess / (double) championMetrics.size()) * 100 : 0.0)
                .challengeSuccessRate(challengeMetrics.size() > 0 ? (challSuccess / (double) challengeMetrics.size()) * 100 : 0.0)
                .build();

        return AnalyticsResponse.builder()
                .summaryCards(summaryCards)
                .executionTimeData(executionTimeData)
                .pieData(pieData)
                .successRateData(successRateData)
                .cumulativeData(cumulativeData)
                .radarData(radarData)
                .performanceComparison(performanceComparison)
                .detailedStatistics(detailedStatistics)
                .build();
    }

    private ExecutionResponse mapToResponse(ChampionChallengeExecutionEntity execution) {
        List<NodeMetricResponse> championMetrics = execution.getNodeMetrics().stream()
                .filter(m -> "CHAMPION".equals(m.getVariant()))
                .map(this::mapMetricToResponse)
                .collect(Collectors.toList());

        List<NodeMetricResponse> challengeMetrics = execution.getNodeMetrics().stream()
                .filter(m -> "CHALLENGE".equals(m.getVariant()))
                .map(this::mapMetricToResponse)
                .collect(Collectors.toList());

        return ExecutionResponse.builder()
                .id(execution.getId())
                .name(execution.getName())
                .description(execution.getDescription())
                .championWorkflowId(execution.getChampionWorkflowId())
                .challengeWorkflowId(execution.getChallengeWorkflowId())
                .requestPayload(execution.getRequestPayload())
                .status(execution.getStatus().name())
                .startedAt(execution.getStartedAt())
                .completedAt(execution.getCompletedAt())
                .createdAt(execution.getCreatedAt())
                .createdBy(execution.getCreatedBy())
                .totalChampionTimeMs(execution.getTotalChampionTimeMs())
                .totalChallengeTimeMs(execution.getTotalChallengeTimeMs())
                .winner(execution.getWinner())
                .championMetrics(championMetrics)
                .challengeMetrics(challengeMetrics)
                .build();
    }

    private NodeMetricResponse mapMetricToResponse(ExecutionNodeMetricEntity metric) {
        return NodeMetricResponse.builder()
                .id(metric.getId())
                .variant(metric.getVariant())
                .nodeId(metric.getNodeId())
                .nodeName(metric.getNodeName())
                .nodeType(metric.getNodeType())
                .requestData(metric.getRequestData())
                .responseData(metric.getResponseData())
                .executionTimeMs(metric.getExecutionTimeMs())
                .status(metric.getStatus())
                .errorMessage(metric.getErrorMessage())
                .startedAt(metric.getStartedAt())
                .completedAt(metric.getCompletedAt())
                .metadata(metric.getMetadata())
                .build();
    }
}
